src/language_features/kak_lsp_custom.rs

```diff
@@ -0,0 +1,318 @@
+use crate::{
+    context::*, language_features::hover::editor_hover, markup::escape_brace, types::*, util::*,
+};
+use lsp_types::{request::*, *};
+use serde::Deserialize;
+use url::Url;
+
+/// Entry point for the `kak-lsp/next-or-previous-symbol` functionality
+///
+/// Essentially all this method is doing is making a DocumentSymbolRequest
+/// and then passing on the results to `process_results_for_next_or_prev()`.
+///
+/// This function's code is derived from `language_features::text_document_document_symbol()`
+/// and similar to it. At this stage we just need to issue a `DocumentSymbol` request.
+pub fn document_next_or_prev_symbol_request(
+    meta: EditorMeta,
+    editor_params: EditorParams,
+    ctx: &mut Context,
+) {
+    let req_params = DocumentSymbolParams {
+        text_document: TextDocumentIdentifier {
+            uri: Url::from_file_path(&meta.buffile).unwrap(),
+        },
+        partial_result_params: Default::default(),
+        work_done_progress_params: Default::default(),
+    };
+    ctx.call::<DocumentSymbolRequest, _>(
+        meta,
+        req_params,
+        move |ctx: &mut Context, meta, result| {
+            process_results_for_next_or_prev(meta, editor_params, result, ctx)
+        },
+    );
+}
+
+/// When this function is called we have got the results of DocumentSymbolRequest.
+/// This function does some minimal processing and then hands the data off to an
+/// appropriate function that will search for the next/previous symbol.
+///
+/// After that is done, we call `next_or_prev_response()` which does the job
+/// of actually sending the response back to Kakoune.
+///
+/// This function's code is derived from `document_symbol::editor_document_symbol()`.
+/// and its basic structure is similar to it.
+fn process_results_for_next_or_prev(
+    meta: EditorMeta,
+    editor_params: EditorParams,
+    result: Option<DocumentSymbolResponse>,
+    ctx: &mut Context,
+) {
+    let params = NextOrPrevSymbolParams::deserialize(editor_params).unwrap();
+    let hover = params.hover;
+    let maybe_details = match result {
+        Some(DocumentSymbolResponse::Flat(mut result)) => {
+            if result.is_empty() {
+                return;
+            }
+            // First let's sort the results so we can find next and previous properly
+            // This step does _not_ happen in `document_symbol::editor_document_symbol()`.
+            //
+            // Some language servers return symbol locations in unsorted order
+            // or non-ascending order.
+            result.sort_by(|a, b| a.location.range.start.cmp(&b.location.range.start));
```

Comment:
The docs say

> When applicable, unstable sorting is preferred because it is generally faster

do we expect symbols that start at the same location? Seems hard to find out

Reply:
The requirement to sort was originally triggered by the `ocamllsp` which was returning its results for document symbol in descending line order instead of ascending order which caused the next / previous logic to get messed up.

So then I just implemented the easiest way to sort things. I didn't think about stable / unstable sorting at that point. 

In general, there are quite a few inefficiencies in the kakoune <-> kak-lsp pipeline with shells being invoked from kakoune quite indiscriminately. In my view we should not worry about stable / unstable sort distinction as everything is in-memory and generally about ~100-200 items. 

You are free to mess around with this in the future, of course, to how you see fit. Lets leave as is for now?

Reply:
P.S. I guess you are just suggesting we use `sort_unstable_by()` right? Its a trivial change, however, I don't have strong views on this. If you feel you want to use the unstable version, you can definitely change in future.

Reply:
<details>

It's not only performance, but about adding extra assumptions for the reader.
If the code follows the simple rule "use unstable sort unless stable sort is required" then a `.sort_unstable()` communicates that we don't get equal elements (or we don't care about their order).
For the same reason I'd rather use `x.unwrap()` than `if let Some(x) = x` if `x` is guaranteed to be filled.

</details>

Ok nevermind, it looks like we actually *can* get two symbols with identical location, using good old C macros.

```c
#define CLEVER_MACRO int x, y;

CLEVER_MACRO;
```

clangd gives me

	test.c:3:1:Variable x
	test.c:3:1:Variable y

so the stable sort is actually important, because otherwise the order might change randomly even if clangd is consistent